File: voice-agents/android_kotlin_voice_widget/app/src/main/java/ai/smallest/atomswidget/AtomsClient.kt
package ai.smallest.atomswidget

import android.util.Log
import kotlinx.coroutines.*
import okhttp3.*
import okio.ByteString
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * OkHttp WebSocket wrapper with reconnect-on-transient + auth-close hard-stop.
 * All callbacks fire on OkHttp's internal executor; do not touch UI state
 * from them — the view-model hops to the main thread via viewModelScope.
 */
class AtomsClient(
    private val apiKey: String,
    private val agentId: String,
    private val sampleRate: Int,
    private val onEvent: (ServerEvent) -> Unit,
    private val onFatalError: (String) -> Unit,
) {
    sealed class ServerEvent {
        data class SessionCreated(val sessionId: String, val callId: String) : ServerEvent()
        data class OutputAudioDelta(val base64: String) : ServerEvent()
        data object AgentStartTalking : ServerEvent()
        data object AgentStopTalking : ServerEvent()
        data object Interruption : ServerEvent()
        data class SessionClosed(val reason: String?) : ServerEvent()
        data class Error(val code: String, val message: String) : ServerEvent()
    }

    // Own the scope: the microphone coroutine is launched asynchronously from
    // OkHttp's onOpen callback, so it must outlive any caller stack frame.
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()
    private var webSocket: WebSocket? = null
    private var micJob: Job? = null
    private var reconnectAttempt = 0
    private val backoffMillis = longArrayOf(500, 1_000, 2_000, 5_000, 15_000)
    private val maxReconnects = 5
    private var explicitlyClosed = false

    private val onOpen = CompletableDeferred<Unit>()

    fun start(startMic: suspend (WebSocket) -> Unit) {
        explicitlyClosed = false
        connect(startMic)
    }

    fun sendMicChunk(base64: String) {
        val ws = webSocket ?: return
        val payload = JSONObject().apply {
            put("type", "input_audio_buffer.append")
            put("audio", base64)
        }
        ws.send(payload.toString())
    }

    fun stop() {
        explicitlyClosed = true
        micJob?.cancel()
        webSocket?.close(1000, "client stop")
        webSocket = null
        scope.cancel()
    }

    private fun connect(startMic: suspend (WebSocket) -> Unit) {
        val url = HttpUrl.Builder()
            .scheme("https")  // OkHttp wraps wss via https
            .host("api.smallest.ai")
            .addPathSegments("atoms/v1/agent/connect")
            .addQueryParameter("token", apiKey)
            .addQueryParameter("agent_id", agentId)
            .addQueryParameter("mode", "webcall")
            .addQueryParameter("sample_rate", sampleRate.toString())
            .build()

        val request = Request.Builder().url(url).build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(ws: WebSocket, response: Response) {
                Log.i("Atoms", "WebSocket open")
                reconnectAttempt = 0
                micJob = scope.launch { startMic(ws) }
            }
            override fun onMessage(ws: WebSocket, text: String) = handleEvent(text)
            override fun onMessage(ws: WebSocket, bytes: ByteString) = handleEvent(bytes.utf8())
            override fun onClosing(ws: WebSocket, code: Int, reason: String) {
                Log.i("Atoms", "WebSocket closing $code $reason")
            }
            override fun onClosed(ws: WebSocket, code: Int, reason: String) {
                Log.i("Atoms", "WebSocket closed $code $reason")
            }
            override fun onFailure(ws: WebSocket, t: Throwable, r: Response?) {
                Log.e("Atoms", "WS failure: ${t.message}")
                if (explicitlyClosed) return
                val authFail = r?.code == 401 || r?.code == 403
                if (authFail || reconnectAttempt >= maxReconnects) {
                    onFatalError(
                        if (authFail) "auth rejected (${r?.code})"
                        else "reconnect gave up after $maxReconnects attempts"
                    )
                    return
                }
                val delay = backoffMillis[minOf(reconnectAttempt, backoffMillis.size - 1)]
                reconnectAttempt += 1
                scope.launch {
                    delay(delay)
                    connect(startMic)
                }
            }
        })
    }

    private fun handleEvent(text: String) {
        val json = try { JSONObject(text) } catch (_: Exception) { return }
        when (json.optString("type")) {
            "session.created" -> onEvent(ServerEvent.SessionCreated(
                json.optString("session_id"), json.optString("call_id")
            ))
            "output_audio.delta" -> onEvent(ServerEvent.OutputAudioDelta(json.optString("audio")))
            "agent_start_talking" -> onEvent(ServerEvent.AgentStartTalking)
            "agent_stop_talking" -> onEvent(ServerEvent.AgentStopTalking)
            "interruption" -> onEvent(ServerEvent.Interruption)
            "session.closed" -> onEvent(ServerEvent.SessionClosed(json.optString("reason").ifEmpty { null }))
            "error" -> onEvent(ServerEvent.Error(json.optString("code"), json.optString("message")))
        }
    }
}

File: voice-agents/react_native_voice_agent/src/agent/AtomsClient.ts
import { ServerEvent, SessionError } from './types';

const CLEAN_CLOSE_CODES = new Set([1000, 4401, 4403]);
const MAX_RECONNECT_ATTEMPTS = 5;
const RECONNECT_BACKOFF_MS = [500, 1000, 2000, 5000, 15000];

export interface AtomsClientConfig {
  apiKey: string;
  agentId: string;
  sampleRate: number;
  onEvent: (event: ServerEvent) => void;
  onOpen: () => void;
  onClose: (wasClean: boolean, code: number, reason: string) => void;
  onFatalError: (error: SessionError) => void;
}

// Wraps WebSocket with exponential-backoff reconnect for transient drops.
// Connects on start(); user must call close() to teardown. Never retries on
// auth failures (4401/4403) or clean closes (1000).
export class AtomsClient {
  private ws: WebSocket | null = null;
  private reconnectAttempt = 0;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private explicitlyClosed = false;
  private config: AtomsClientConfig;

  constructor(config: AtomsClientConfig) {
    this.config = config;
  }

  start(): void {
    this.explicitlyClosed = false;
    this.connect();
  }

  sendMicChunk(base64Int16LE: string): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    this.ws.send(JSON.stringify({
      type: 'input_audio_buffer.append',
      audio: base64Int16LE,
    }));
  }

  close(reason: string = 'client end'): void {
    this.explicitlyClosed = true;
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.close(1000, reason);
    }
    this.ws = null;
  }

  private connect(): void {
    const { apiKey, agentId, sampleRate } = this.config;
    const url =
      'wss://api.smallest.ai/atoms/v1/agent/connect' +
      `?token=${encodeURIComponent(apiKey)}` +
      `&agent_id=${encodeURIComponent(agentId)}` +
      `&mode=webcall&sample_rate=${sampleRate}`;

    let ws: WebSocket;
    try {
      ws = new WebSocket(url);
    } catch (e) {
      this.config.onFatalError({
        kind: 'network',
        message: e instanceof Error ? e.message : 'websocket construction failed',
        retryable: true,
      });
      return;
    }
    this.ws = ws;

    ws.onopen = () => {
      this.reconnectAttempt = 0;
      this.config.onOpen();
    };

    ws.onmessage = (event) => {
      if (typeof event.data !== 'string') return;
      try {
        const parsed = JSON.parse(event.data) as ServerEvent;
        this.config.onEvent(parsed);
      } catch {
        // malformed JSON from server is non-fatal; drop the frame
      }
    };

    ws.onerror = () => {
      // onclose will fire right after; defer handling there
    };

    ws.onclose = (event) => {
      const wasClean = CLEAN_CLOSE_CODES.has(event.code);
      this.config.onClose(wasClean, event.code, event.reason || '');

      if (this.explicitlyClosed || wasClean) return;

      if (event.code === 4401 || event.code === 4403) {
        this.config.onFatalError({
          kind: 'auth',
          message: `auth rejected (${event.code})`,
          retryable: false,
        });
        return;
      }

      if (this.reconnectAttempt >= MAX_RECONNECT_ATTEMPTS) {
        this.config.onFatalError({
          kind: 'network',
          message: `reconnect gave up after ${MAX_RECONNECT_ATTEMPTS} attempts`,
          retryable: true,
        });
        return;
      }

      const delay = RECONNECT_BACKOFF_MS[this.reconnectAttempt] ?? 15000;
      this.reconnectAttempt += 1;
      this.reconnectTimer = setTimeout(() => this.connect(), delay);
    };
  }
}
